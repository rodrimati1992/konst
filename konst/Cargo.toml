[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

