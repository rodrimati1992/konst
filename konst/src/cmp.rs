//! Comparisong-related items.

mod cmp_wrapper;

mod const_cmp;

pub use self::{
    cmp_wrapper::CmpWrapper,
    const_cmp::{ConstCmp, ConstCmpUnref, IsAConstCmp},
};

#[doc(no_inline)]
pub use crate::polymorphism::kinds::{IsNotStdKind, IsRefKind, IsStdKind};
