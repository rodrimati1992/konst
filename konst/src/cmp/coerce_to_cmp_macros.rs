/// Coerces `reference` to a type that has a `const_eq` or `const_cmp` method.
///
/// # Behavior
///
/// This requires arguments to implement the [`ConstCmp`] trait.
///
/// When a type from the standard library is passed,
/// this wraps it inside a [`CmpWrapper`],
/// which declares `const_eq` and `const_cmp` methods for many standard library types.
///
/// When a user-defined type is used, this evaluates to a reference to the passed in value,
/// dereferencing it as necessary.
///
/// # Limitations
///
/// The parameter(s) must be concrete types, and have a fully inferred type.
/// eg: if you pass an integer literal it must have a suffix to indicate its type.
///
/// # Example
///
/// ```rust
/// use konst::cmp::{CmpWrapper, coerce_to_cmp, impl_cmp};
///
/// struct Unit;
///
/// impl_cmp!{
///     impl Unit;
///     
///     pub const fn const_eq(&self, other: &Self) -> bool {
///         true
///     }
/// }
///
/// let wrapper: CmpWrapper<i32> = coerce_to_cmp!(0i32);
/// assert!( wrapper.const_eq(&0));
/// assert!(!wrapper.const_eq(&1));
///
/// let unit: &Unit = coerce_to_cmp!(Unit);
/// assert!( unit.const_eq(&Unit));
///
///
///
///
///
///
/// ```
///
/// [`ConstCmp`]: crate::cmp::ConstCmp
/// [`CmpWrapper`]: crate::cmp::CmpWrapper
///
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub use crate::__coerce_to_cmp as coerce_to_cmp;

#[doc(hidden)]
#[macro_export]
macro_rules! __coerce_to_cmp {
    ($reference:expr $(,)*) => {{
        match $reference {
            ref reference => {
                let marker = $crate::__::IsAConstCmp::NEW;
                if false {
                    _ = marker.infer_type(reference);
                }
                marker.coerce(reference)
            }
        }
    }};
}

#[doc(hidden)]
#[macro_export]
macro_rules! __coerce_to_cmp2 {
    ($left:expr, $right:expr $(,)*) => {{
        match ($left, $right) {
            (left, right) => {
                let l_marker = $crate::__::IsAConstCmp::NEW;
                let r_marker = $crate::__::IsAConstCmp::NEW;
                if false {
                    _ = l_marker.infer_type(left);
                    _ = r_marker.infer_type(right);
                }
                (l_marker.coerce(left), r_marker.unreference(right))
            }
        }
    }};
}
