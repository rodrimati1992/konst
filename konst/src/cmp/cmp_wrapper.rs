use crate::cmp::{ConstCmp, IsAConstCmp, IsNotStdKind, IsStdKind};

/// A wrapper type for std types, which defines `const_eq` and `const_cmp` methods for them.
///
/// This is what [`coerce_to_cmp`](crate::coerce_to_cmp)
/// and the comparison macros convert standard library types into.
///
/// # Example
///
/// ```rust
/// use konst::{
///     cmp::CmpWrapper,
///     coerce_to_cmp,
/// };
///
/// use std::cmp::Ordering;
///
/// {
///     // The `CmpWrapper<u32>` type annotation is just for the reader
///     let foo: CmpWrapper<u32> = coerce_to_cmp!(10u32);
///     assert!( foo.const_eq(&10));
///     assert!(!foo.const_eq(&20));
///     
///     assert_eq!(foo.const_cmp(&5), Ordering::Greater);
///     assert_eq!(foo.const_cmp(&10), Ordering::Equal);
///     assert_eq!(foo.const_cmp(&15), Ordering::Less);
/// }
/// {
///     let bar = CmpWrapper(Ordering::Equal);
///     assert!( bar.const_eq(&Ordering::Equal));
///     assert!(!bar.const_eq(&Ordering::Less));
///     
///     assert_eq!(bar.const_cmp(&Ordering::Less), Ordering::Greater);
///     assert_eq!(bar.const_cmp(&Ordering::Equal), Ordering::Equal);
///     assert_eq!(bar.const_cmp(&Ordering::Greater), Ordering::Less);
/// }
///
/// ```
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct CmpWrapper<T>(pub T);

impl<'a, T> CmpWrapper<&'a [T]> {
    /// For constructing from a reference to an array.
    ///
    /// With slices you can do `CmpWrapper(slice)` as well.
    #[inline(always)]
    pub const fn slice(x: &'a [T]) -> Self {
        Self { 0: x }
    }
}

impl<P> ConstCmp for CmpWrapper<P> {
    type Kind = IsNotStdKind;
}

macro_rules! std_kind_impls {
    ($($ty:ty),* $(,)* ) => (
        $(
            impl ConstCmp for $ty {
                type Kind = IsStdKind;
            }

            impl<T> IsAConstCmp<IsStdKind, $ty, T> {
                /// Copies the value from `reference`, and wraps it in a `CmpWrapper`
                #[inline(always)]
                pub const fn coerce(self, reference: &$ty) -> CmpWrapper<$ty> {
                    CmpWrapper(*reference)
                }
            }

            impl CmpWrapper<$ty> {
                /// Compares `self` and `other` for equality.
                #[inline(always)]
                pub const fn const_eq(self, other: &$ty) -> bool {
                    self.0 == *other
                }
            }
        )*
    )
}

std_kind_impls! {
    i8, u8,
    i16, u16,
    i32, u32,
    i64, u64,
    i128, u128,
    isize, usize,
    bool, char,
}
