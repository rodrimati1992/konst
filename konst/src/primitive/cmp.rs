use core::cmp::Ordering;

macro_rules! declare_int_cmp_fn {
    (
        ($type:ty, ($cmp_fn_name:ident))
        docs( $docs_eq:expr, $docs_cmp:expr, )
    ) => {
        crate::__delegate_const_ord! {
            #[doc = $docs_cmp]
            pub const fn $cmp_fn_name(copy left: $type, right: $type) -> Ordering {
                cmp_int!(left, right)
            }
        }
    };
}

__declare_fns_with_docs! {
    (u8, (cmp_u8))
    (u16, (cmp_u16))
    (u32, (cmp_u32))
    (u64, (cmp_u64))
    (u128, (cmp_u128))
    (usize, (cmp_usize))

    (i8, (cmp_i8))
    (i16, (cmp_i16))
    (i32, (cmp_i32))
    (i64, (cmp_i64))
    (i128, (cmp_i128))
    (isize, (cmp_isize))

    (bool, (cmp_bool))
    (char, (cmp_char))

    docs(default)

    macro = declare_int_cmp_fn!(),
}

__declare_fns_with_docs! {
    (Option<u8>, (eq_option_u8, cmp_option_u8))
    (Option<i8>, (eq_option_i8, cmp_option_i8))
    (Option<u16>, (eq_option_u16, cmp_option_u16))
    (Option<i16>, (eq_option_i16, cmp_option_i16))
    (Option<u32>, (eq_option_u32, cmp_option_u32))
    (Option<i32>, (eq_option_i32, cmp_option_i32))
    (Option<u64>, (eq_option_u64, cmp_option_u64))
    (Option<i64>, (eq_option_i64, cmp_option_i64))
    (Option<u128>, (eq_option_u128, cmp_option_u128))
    (Option<i128>, (eq_option_i128, cmp_option_i128))
    (Option<usize>, (eq_option_usize, cmp_option_usize))
    (Option<isize>, (eq_option_isize, cmp_option_isize))
    (Option<bool>, (eq_option_bool, cmp_option_bool))
    (Option<char>, (eq_option_char, cmp_option_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        params(l, r)
        eq_comparison = l == r,
        cmp_comparison = cmp_int!(l, r),
        parameter_copyability = copy,
    ),
}
