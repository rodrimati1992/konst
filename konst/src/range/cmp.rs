use core::ops::{Range, RangeInclusive};

use crate::cmp::{CmpWrapper, ConstCmp, IsAConstCmp, IsStdKind};

macro_rules! shared_range_impls {
    ($type:ty, $eq_fn_name:ident) => {
        impl ConstCmp for $type {
            type Kind = IsStdKind;
        }
        impl CmpWrapper<$type> {
            /// Compares `self` and `other` for equality.
            #[inline(always)]
            pub const fn const_eq(&self, other: &$type) -> bool {
                $eq_fn_name(&self.0, other)
            }
        }
    };
}

macro_rules! declare_range_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(Range {
                    start: range.start,
                    end: range.end,
                })
            }
        }

        __delegate_const_eq! {
            skip_coerce;
            for['a,]

            #[doc = $docs_eq]
            pub const fn $eq_fn_name(ref left: &'a $type, right: &'a $type) -> bool {
                left.start == right.start && left.end == right.end
            }
        }
    };
}

__declare_fns_with_docs! {
    (Range<u8>,    (eq_range_u8))
    (Range<u16>,   (eq_range_u16))
    (Range<u32>,   (eq_range_u32))
    (Range<u64>,   (eq_range_u64))
    (Range<u128>,  (eq_range_u128))
    (Range<usize>, (eq_range_usize))

    (Range<char>, (eq_range_char))

    docs(default)

    macro = declare_range_cmp_fns!(),
}

macro_rules! declare_rangeinclusive_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(RangeInclusive::new(*range.start(), *range.end()))
            }
        }

        #[doc = $docs_eq]
        pub const fn $eq_fn_name(left: &$type, right: &$type) -> bool {
            *left.start() == *right.start() && *left.end() == *right.end()
        }
    };
}

__declare_fns_with_docs! {
    (RangeInclusive<u8>,    (eq_rangeinc_u8,))
    (RangeInclusive<u16>,   (eq_rangeinc_u16))
    (RangeInclusive<u32>,   (eq_rangeinc_u32))
    (RangeInclusive<u64>,   (eq_rangeinc_u64))
    (RangeInclusive<u128>,  (eq_rangeinc_u128))
    (RangeInclusive<usize>, (eq_rangeinc_usize))

    (RangeInclusive<char>, (eq_rangeinc_char))

    docs(default)

    macro = declare_rangeinclusive_cmp_fns!(),
}
