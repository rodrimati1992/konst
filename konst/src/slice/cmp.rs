#[doc(inline)]
pub use crate::slice::{
    cmp_bytes as cmp_slice_u8, cmp_option_bytes as cmp_option_slice_u8, eq_bytes as eq_slice_u8,
    eq_option_bytes as eq_option_slice_u8,
};

use core::cmp::Ordering;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (,,, u16, eq_slice_u16, cmp_slice_u16,)
    (,,, u32, eq_slice_u32, cmp_slice_u32,)
    (,,, u64, eq_slice_u64, cmp_slice_u64,)
    (,,, u128, eq_slice_u128, cmp_slice_u128,)
    (,,, usize, eq_slice_usize, cmp_slice_usize,)

    (,,, i8, eq_slice_i8, cmp_slice_i8,)
    (,,, i16, eq_slice_i16, cmp_slice_i16,)
    (,,, i32, eq_slice_i32, cmp_slice_i32,)
    (,,, i64, eq_slice_i64, cmp_slice_i64,)
    (,,, i128, eq_slice_i128, cmp_slice_i128,)
    (,,, isize, eq_slice_isize, cmp_slice_isize,)

    (,,, bool, eq_slice_bool, cmp_slice_bool,)
    (,,, char, eq_slice_char, cmp_slice_char,)
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&str]` for equality.
    pub const fn eq_slice_str(ref l: &[&str], r: &[&str]) -> bool {
        crate::const_eq_for!(slice; l, r, crate::eq_str)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&str]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_str(ref left: &[&str], right: &[&str]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, crate::cmp_str)
    }
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&[u8]]` for equality.
    pub const fn eq_slice_bytes(ref l: &[&[u8]], r: &[&[u8]]) -> bool {
        crate::const_eq_for!(slice; l, r, eq_slice_u8)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&[u8]]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_bytes(ref left: &[&[u8]], right: &[&[u8]]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, cmp_slice_u8)
    }
}

__declare_fns_with_docs! {
    (Option<&'a [u16]>, (eq_option_slice_u16, cmp_option_slice_u16))
    (Option<&'a [u32]>, (eq_option_slice_u32, cmp_option_slice_u32))
    (Option<&'a [u64]>, (eq_option_slice_u64, cmp_option_slice_u64))
    (Option<&'a [u128]>, (eq_option_slice_u128, cmp_option_slice_u128))
    (Option<&'a [usize]>, (eq_option_slice_usize, cmp_option_slice_usize))
    (Option<&'a [i8]>, (eq_option_slice_i8, cmp_option_slice_i8))
    (Option<&'a [i16]>, (eq_option_slice_i16, cmp_option_slice_i16))
    (Option<&'a [i32]>, (eq_option_slice_i32, cmp_option_slice_i32))
    (Option<&'a [i64]>, (eq_option_slice_i64, cmp_option_slice_i64))
    (Option<&'a [i128]>, (eq_option_slice_i128, cmp_option_slice_i128))
    (Option<&'a [isize]>, (eq_option_slice_isize, cmp_option_slice_isize))
    (Option<&'a [bool]>, (eq_option_slice_bool, cmp_option_slice_bool))
    (Option<&'a [char]>, (eq_option_slice_char, cmp_option_slice_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}

__declare_fns_with_docs! {
    (Option<&'a [&'b str]>, (eq_option_slice_str, cmp_option_slice_str))
    (Option<&'a [&'b [u8]]>, (eq_option_slice_bytes, cmp_option_slice_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a, 'b,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}
