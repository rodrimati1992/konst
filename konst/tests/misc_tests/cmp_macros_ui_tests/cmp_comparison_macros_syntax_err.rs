use konst::cmp;

use core::cmp::Ordering;


const _: bool = cmp::const_eq!();
const _: bool = cmp::const_eq!(0u8);
const _: bool = cmp::const_eq!(0u8,);

const _: bool = cmp::const_ne!();
const _: bool = cmp::const_ne!(0u8);
const _: bool = cmp::const_ne!(0u8,);

const _: Ordering = cmp::const_cmp!();
const _: Ordering = cmp::const_cmp!(0u8);
const _: Ordering = cmp::const_cmp!(0u8, );

const _: bool = cmp::const_lt!();
const _: bool = cmp::const_lt!(0u8);
const _: bool = cmp::const_lt!(0u8,);

const _: bool = cmp::const_le!();
const _: bool = cmp::const_le!(0u8);
const _: bool = cmp::const_le!(0u8,);

const _: bool = cmp::const_gt!();
const _: bool = cmp::const_gt!(0u8);
const _: bool = cmp::const_gt!(0u8,);

const _: bool = cmp::const_ge!();
const _: bool = cmp::const_ge!(0u8);
const _: bool = cmp::const_ge!(0u8,);



fn main() {}