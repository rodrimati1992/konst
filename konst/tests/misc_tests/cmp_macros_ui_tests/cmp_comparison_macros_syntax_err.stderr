error: unexpected end of macro invocation
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:6:17
  |
6 | const _: bool = cmp::const_eq!();
  |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
note: while trying to match meta-variable `$left:expr`
 --> src/cmp/const_eq_macros.rs
  |
  |     ($left:expr, $right:expr $(,)*) => {
  |      ^^^^^^^^^^

error: unexpected end of macro invocation
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:7:35
  |
7 | const _: bool = cmp::const_eq!(0u8);
  |                                   ^ missing tokens in macro arguments
  |
note: while trying to match `,`
 --> src/cmp/const_eq_macros.rs
  |
  |     ($left:expr, $right:expr $(,)*) => {
  |                ^

error: unexpected end of macro invocation
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:8:36
  |
8 | const _: bool = cmp::const_eq!(0u8,);
  |                                    ^ missing tokens in macro arguments
  |
note: while trying to match meta-variable `$right:expr`
 --> src/cmp/const_eq_macros.rs
  |
  |     ($left:expr, $right:expr $(,)*) => {
  |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:10:17
   |
10 | const _: bool = cmp::const_ne!();
   |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^
   = note: this error originates in the macro `cmp::const_ne` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:11:35
   |
11 | const _: bool = cmp::const_ne!(0u8);
   |                                   ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:12:36
   |
12 | const _: bool = cmp::const_ne!(0u8,);
   |                                    ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:14:21
   |
14 | const _: Ordering = cmp::const_cmp!();
   |                     ^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:15:40
   |
15 | const _: Ordering = cmp::const_cmp!(0u8);
   |                                        ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:16:41
   |
16 | const _: Ordering = cmp::const_cmp!(0u8, );
   |                                         ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:18:17
   |
18 | const _: bool = cmp::const_lt!();
   |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:19:35
   |
19 | const _: bool = cmp::const_lt!(0u8);
   |                                   ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:20:36
   |
20 | const _: bool = cmp::const_lt!(0u8,);
   |                                    ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:22:17
   |
22 | const _: bool = cmp::const_le!();
   |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:23:35
   |
23 | const _: bool = cmp::const_le!(0u8);
   |                                   ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:24:36
   |
24 | const _: bool = cmp::const_le!(0u8,);
   |                                    ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:26:17
   |
26 | const _: bool = cmp::const_gt!();
   |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:27:35
   |
27 | const _: bool = cmp::const_gt!(0u8);
   |                                   ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:28:36
   |
28 | const _: bool = cmp::const_gt!(0u8,);
   |                                    ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:30:17
   |
30 | const _: bool = cmp::const_ge!();
   |                 ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:31:35
   |
31 | const _: bool = cmp::const_ge!(0u8);
   |                                   ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_macros_syntax_err.rs:32:36
   |
32 | const _: bool = cmp::const_ge!(0u8,);
   |                                    ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/cmp/const_ltgt_macros.rs
   |
   |     ($left:expr, $right:expr $(,)*) => {
   |                  ^^^^^^^^^^^
